### DISCUSSION TITLE
How to use `IEnumerable<Tuple<...>>` in xUnit v3?

### DISCUSSION THREAD

**Discussion by glen-84** [xunit/xunit#3100] [Question] [answered]
The [What’s New in v3](https://xunit.net/docs/getting-started/v3/whats-new) guide mentions:

> In v3, this contract has been expanded to allow three different legal data row representations:
> ...
> named or unnamed Tuple<>

But I can't find an example of what this looks like, and it's not working for me.

```cs
public static IEnumerable<Tuple<string[], string[]>> MemberData()
```

> MemberData must reference a data type assignable to 'System.Collections.Generic.IEnumerable<object[]>', 'System.Collections.Generic.IAsyncEnumerable<object[]>', 'System.Collections.Generic.IEnumerable<Xunit.ITheoryDataRow>', or 'System.Collections.Generic.IAsyncEnumerable<Xunit.ITheoryDataRow>'. The referenced type 'System.Collections.Generic.IEnumerable<System.Tuple<string[], string[]>>' is not valid.

```cs
public static IEnumerable<(string[] A, string[] B)> MemberData()
```

> MemberData must reference a data type assignable to 'System.Collections.Generic.IEnumerable<object[]>', 'System.Collections.Generic.IAsyncEnumerable<object[]>', 'System.Collections.Generic.IEnumerable<Xunit.ITheoryDataRow>', or 'System.Collections.Generic.IAsyncEnumerable<Xunit.ITheoryDataRow>'. The referenced type 'System.Collections.Generic.IEnumerable<(string[] Sdl, string[] ErrorMessages)>' is not valid.

What am I missing here?

**Comment by bradwilson** [comment, marked as answer]
It does work, but it appears I haven't updated xUnit1019 yet to add that as allowable.

Using a suppression to turn it off temporarily, this test does pass:

```csharp
public class UnitTest
{
    public static IEnumerable<(string[], string[])> DataSource =>
    [
        (["a", "b", "c"], ["c", "a", "b"])
    ];

    [Theory]
#pragma warning disable xUnit1019
    [MemberData(nameof(DataSource))]
#pragma warning restore xUnit1019
    public void TestMethod(string[] a, string[] b) =>
        Assert.Equivalent(a, b);
}
```

I will create an issue to get xUnit1019 updated for this. Thanks!

**Reply by bradwilson** [reply]
Bug opened here: https://github.com/xunit/xunit/issues/3102

**Reply by bradwilson** [reply]
Fix available in `xunit.analyzers` package `1.19.0-pre.12` https://xunit.net/docs/using-ci-builds
